File: backend-rust/src/jobs.rs
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::Instant;

use crate::types::Track;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum JobStatus {
    Processing,
    Completed,
    Failed,
}

impl JobStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            JobStatus::Processing => "processing",
            JobStatus::Completed => "completed",
            JobStatus::Failed => "failed",
        }
    }
}

#[derive(Debug, Clone)]
pub struct JobSnapshot {
    pub job_id: String,
    pub status: JobStatus,
    pub frame_count: i32,
    pub expected_frames: i32,
    pub started_at: Instant,
    pub finished_at: Option<Instant>,
    pub tracks: Option<Vec<Track>>,
    pub video_id: Option<String>,
    pub error: Option<String>,
}

struct Job {
    job_id: String,
    status: JobStatus,
    frame_count: i32,
    expected_frames: i32,
    total_detections: i32,
    started_at: Instant,
    finished_at: Option<Instant>,
    tracks: Option<Vec<Track>>,
    video_id: Option<String>,
    error: Option<String>,
}

impl Job {
    fn snapshot(&self) -> JobSnapshot {
        JobSnapshot {
            job_id: self.job_id.clone(),
            status: self.status,
            frame_count: self.frame_count,
            expected_frames: self.expected_frames,
            started_at: self.started_at,
            finished_at: self.finished_at,
            tracks: self.tracks.clone(),
            video_id: self.video_id.clone(),
            error: self.error.clone(),
        }
    }
}

/// In-memory job tracking, mirroring the Python `app/core/video_jobs.py`.
pub struct JobStore {
    jobs: Mutex<HashMap<String, Job>>,
}

impl JobStore {
    pub fn new() -> Self {
        Self {
            jobs: Mutex::new(HashMap::new()),
        }
    }

    pub fn create(&self, job_id: String) {
        let mut jobs = self.jobs.lock().unwrap();
        jobs.insert(
            job_id.clone(),
            Job {
                job_id,
                status: JobStatus::Processing,
                frame_count: 0,
                expected_frames: 0,
                total_detections: 0,
                started_at: Instant::now(),
                finished_at: None,
                tracks: None,
                video_id: None,
                error: None,
            },
        );
    }

    pub fn update_progress(&self, job_id: &str, frame_count: i32, expected_frames: i32, total_detections: i32) {
        let mut jobs = self.jobs.lock().unwrap();
        if let Some(job) = jobs.get_mut(job_id) {
            job.frame_count = frame_count;
            job.expected_frames = job.expected_frames.max(expected_frames);
            job.total_detections = total_detections;
        }
    }

    pub fn complete(&self, job_id: &str, tracks: Vec<Track>, video_id: Option<String>) {
        let mut jobs = self.jobs.lock().unwrap();
        if let Some(job) = jobs.get_mut(job_id) {
            job.status = JobStatus::Completed;
            job.tracks = Some(tracks);
            job.video_id = video_id;
            job.finished_at = Some(Instant::now());
        }
    }

    pub fn fail(&self, job_id: &str, error: String) {
        let mut jobs = self.jobs.lock().unwrap();
        if let Some(job) = jobs.get_mut(job_id) {
            job.status = JobStatus::Failed;
            job.error = Some(error);
            job.finished_at = Some(Instant::now());
        }
    }

    pub fn get(&self, job_id: &str) -> Option<JobSnapshot> {
        let jobs = self.jobs.lock().unwrap();
        jobs.get(job_id).map(Job::snapshot)
    }
}

File: backend-go/internal/jobs/jobs.go
// Package jobs is a small in-memory tracker for long-running background
// work that a client starts and then polls -- video processing being the
// first user. It knows nothing about what the work actually does; the
// result type is generic so this package stays reusable for any future
// "start it, poll it" endpoint.
package jobs

import (
	"sync"
	"time"
)

type Status string

const (
	StatusProcessing Status = "processing"
	StatusCompleted  Status = "completed"
	StatusFailed     Status = "failed"
)

type job[T any] struct {
	mu             sync.Mutex
	status         Status
	frameCount     int
	expectedFrames int
	startedAt      time.Time
	finishedAt     time.Time
	result         T
	err            error
}

// Store holds every job for the life of the process -- there is no
// persistence and no eviction, so jobs disappear on restart and accumulate
// in memory for long-lived processes. Fine for now; worth revisiting if this
// ever needs to survive a restart or run for weeks without one.
type Store[T any] struct {
	mu   sync.RWMutex
	jobs map[string]*job[T]
}

func NewStore[T any]() *Store[T] {
	return &Store[T]{jobs: make(map[string]*job[T])}
}

func (s *Store[T]) Create(id string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.jobs[id] = &job[T]{status: StatusProcessing, startedAt: time.Now()}
}

func (s *Store[T]) UpdateProgress(id string, frameCount, expectedFrames int) {
	j, ok := s.get(id)
	if !ok {
		return
	}
	j.mu.Lock()
	defer j.mu.Unlock()
	j.frameCount = frameCount
	j.expectedFrames = expectedFrames
}

func (s *Store[T]) Complete(id string, result T) {
	j, ok := s.get(id)
	if !ok {
		return
	}
	j.mu.Lock()
	defer j.mu.Unlock()
	j.status = StatusCompleted
	j.result = result
	j.finishedAt = time.Now()
}

func (s *Store[T]) Fail(id string, err error) {
	j, ok := s.get(id)
	if !ok {
		return
	}
	j.mu.Lock()
	defer j.mu.Unlock()
	j.status = StatusFailed
	j.err = err
	j.finishedAt = time.Now()
}

func (s *Store[T]) get(id string) (*job[T], bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	j, ok := s.jobs[id]
	return j, ok
}

// Snapshot is a point-in-time, immutable copy of a job's state -- safe to
// hold onto and read after the lock is released.
type Snapshot[T any] struct {
	Status         Status
	FrameCount     int
	ExpectedFrames int
	Elapsed        time.Duration
	ETA            *time.Duration
	Err            error
	Result         T
}

func (s *Store[T]) Snapshot(id string) (Snapshot[T], bool) {
	j, ok := s.get(id)
	if !ok {
		return Snapshot[T]{}, false
	}
	j.mu.Lock()
	defer j.mu.Unlock()

	end := time.Now()
	if !j.finishedAt.IsZero() {
		end = j.finishedAt
	}

	snap := Snapshot[T]{
		Status:         j.status,
		FrameCount:     j.frameCount,
		ExpectedFrames: j.expectedFrames,
		Elapsed:        end.Sub(j.startedAt),
		Err:            j.err,
		Result:         j.result,
	}

	if j.status == StatusProcessing && j.frameCount > 0 && j.expectedFrames > j.frameCount {
		rate := snap.Elapsed.Seconds() / float64(j.frameCount)
		remaining := time.Duration(rate*float64(j.expectedFrames-j.frameCount)) * time.Second
		snap.ETA = &remaining
	}

	return snap, true
}
